File: problems/0025-reverse-nodes-in-k-group/solution.go
// Link: https://leetcode.cn/problems/reverse-nodes-in-k-group/
// Level: Hard
// Date: 2025-04-14

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverse(head, tail *ListNode) (*ListNode, *ListNode) {
    prev := tail.Next
    p := head
    for prev != tail {
        nex := p.Next
        p.Next = prev
        prev = p
        p = nex
    }
    return tail, head
}

func reverseKGroup(head *ListNode, k int) *ListNode {
    hair := &ListNode{Next: head}
    pre := hair

    for head != nil {
        tail := pre
        for i := 0; i < k; i++ {
            tail = tail.Next
            if tail == nil {
                return hair.Next
            }
        }
        nex := tail.Next
        head, tail = reverse(head, tail)
        pre.Next = head
        tail.Next = nex
        pre = tail
        head = tail.Next
    }
    return hair.Next
}

File: problems/0025-reverse-nodes-in-k-group/solution.cpp
// Link: https://leetcode.cn/problems/reverse-nodes-in-k-group/
// Level: Hard
// Date: 2025-04-14

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    public:
        pair<ListNode*, ListNode*> reverse(ListNode* head, ListNode* tail) {
            ListNode* prev = tail->next;
            ListNode* p = head;
            while (prev != tail) {
                ListNode* nex = p->next;
                p->next = prev;
                prev = p;
                p = nex;
            }
            return {tail, head};
        }
        ListNode* reverseKGroup(ListNode* head, int k) {
            ListNode* hair = new ListNode();
            hair->next = head;
            ListNode* pre = hair;

            while (head) {
                ListNode* tail = pre;

                for (int i = 0; i < k; i++) {
                    tail = tail->next;
                    if (!tail) {
                        return hair->next;
                    }
                }

                ListNode* nex = tail->next;
                pair<ListNode*, ListNode*> result = reverse(head, tail);
                head = result.first;
                tail = result.second;
                pre->next = head;
                tail->next = nex;
                pre = tail;
                head = tail->next;
            }
            return hair->next;
        }
};